src/Microsoft.Azure.ServiceBus/Core/Batch.cs

```diff
@@ -0,0 +1,144 @@
+        private (string messageId, string sessionId, string partitionKey, string viaPartitionKey) originalMessageData;
+
+        /// <summary>
+        /// Construct a new batch with a maximum batch size and outgoing plugins callback.
+        /// <remarks>
+        /// To construct a batch at run-time, use <see cref="MessageSender"/>, <see cref="QueueClient"/>, or <see cref="TopicClient"/>.
+        /// Use this constructor for testing and custom implementations.
+        /// </remarks>
+        /// </summary>
+        /// <param name="maximumBatchSize">Maximum batch size allowed for batch.</param>
+        /// <param name="pluginsCallback">Plugins callback to invoke on outgoing messages regisered with batch.</param>
+        public Batch(ulong maximumBatchSize, Func<Message, Task<Message>> pluginsCallback)
+        {
+            this.maximumBatchSize = maximumBatchSize;
+            this.pluginsCallback = pluginsCallback;
+            this.datas = new List<Data>();
+            this.result = AmqpMessage.Create(datas);
+        }
+
+        /// <summary>
+        /// Add <see cref="Message"/> to the batch if the overall size of the batch with the added message is not exceeding the batch maximum.
+        /// </summary>
+        /// <param name="message"><see cref="Message"/> to add to the batch.</param>
+        /// <returns></returns>
+        public async Task<bool> TryAdd(Message message)
+        {
+            ThrowIfDisposed();
+
+            message.VerifyMessageIsNotPreviouslyReceived();
+
+            var processedMessage = await pluginsCallback(message);
+
+            var amqpMessage = AmqpMessageConverter.SBMessageToAmqpMessage(processedMessage);
+
+            if (firstMessage == null)
+            {
+                originalMessageData = (processedMessage.MessageId, processedMessage.SessionId, processedMessage.PartitionKey, processedMessage.ViaPartitionKey);
+                firstMessage = amqpMessage;
+            }
+
+            var data = AmqpMessageConverter.ToData(amqpMessage);
+            datas.Add(data);
+
+            if (Size <= maximumBatchSize)
+            {
+                return true;
+            }
+
+            datas.Remove(data);
+            return false;
+
+        }
+
+        /// <summary>
+        /// Number of messages in batch.
+        /// </summary>
+        public int Length => datas.Count;
+
+        internal ulong Size => (ulong) result.SerializedMessageSize;
+
+
+        /// <summary>
+        /// Convert batch to AMQP message.
+        /// </summary>
+        /// <returns></returns>
+        public AmqpMessage ToAmqpMessage()
+        {
+            ThrowIfDisposed();
+
+            if (datas.Count == 1)
+            {
+                firstMessage.Batchable = true;
+                return firstMessage;
+            }
+
+            if (originalMessageData.messageId != null)
+            {
+                result.Properties.MessageId = originalMessageData.messageId;
```

Comment:
>result.Properties.MessageId = originalMessageData.messageId; [](start = 16, length = 60)

What if `TryAdd` succeeded, and then you tried to copy over `MessageId` and other properties, and you crossed your size limit?

Reply:
Absolutely. It also removes the need to "remember" original message system properties and tuple that stores those. Nice!